### ISSUE TITLE
ClickHouse field tables need a TTL

### ISSUE THREAD

**Issue by bnaecker** [oxidecomputer/omicron#6328]
In #3553, we added a TTL of 30 days to the **measurement** tables in ClickHouse. However, the **field** tables do not have a TTL:

```
oxz_clickhouse_aa646c82-c6d7-4d0c-8401-150130927759.local :) show create table fields_uuid;

SHOW CREATE TABLE fields_uuid

Query id: 732d6209-b41c-49c1-9646-713bdb730816

┌─statement──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE oximeter.fields_uuid
(
    `timeseries_name` String,
    `timeseries_key` UInt64,
    `field_name` String,
    `field_value` UUID
)
ENGINE = ReplacingMergeTree
ORDER BY (timeseries_name, field_name, field_value, timeseries_key)
SETTINGS index_granularity = 8192 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.006 sec.
```

This has led to quite a bit of bloat in some of these tables, especially those, like `fields_uuid`, where the field values are quite random and relatively short-lived. For example:

```
oxz_clickhouse_aa646c82-c6d7-4d0c-8401-150130927759.local :) with (select countDistinct(timeseries_key) from measurements_cumulativei64 where timeseries_name = 'crucible_upstairs:read') as unique_measurements, (select countDistinct(timeseries_key) from fields_uuid where timeseries_name = 'crucible_upstairs:read') as unique_fields select unique_measurements, unique_fields, (unique_measurements / unique_fields) as live_fraction;

WITH
    (
        SELECT countDistinct(timeseries_key)
        FROM measurements_cumulativei64
        WHERE timeseries_name = 'crucible_upstairs:read'
    ) AS unique_measurements,
    (
        SELECT countDistinct(timeseries_key)
        FROM fields_uuid
        WHERE timeseries_name = 'crucible_upstairs:read'
    ) AS unique_fields
SELECT
    unique_measurements,
    unique_fields,
    unique_measurements / unique_fields AS live_fraction

Query id: 0a7d0a7f-c0a2-42d1-b920-a1cf698a5304

┌─unique_measurements─┬─unique_fields─┬───────live_fraction─┐
│                1349 │         22865 │ 0.05899846927618631 │
└─────────────────────┴───────────────┴─────────────────────┘

1 row in set. Elapsed: 0.103 sec. Processed 6.06 million rows, 236.97 MB (59.01 million rows/s., 2.31 GB/s.)
Peak memory usage: 13.14 MiB.
```

This shows that we have around 22K unique timeseries today from the `crucible_upstairs:read` schema, while only about 1.3K have measurements that have not been expired by the TTL on the related measurements table. That's an "efficiency" of about 6%, meaning we're storing about 20 times more fields than we need to be.

We should give these tables a haircut, and add a TTL to them. One wrinkle is that these are `ReplacingMergeTree()` tables. We should make sure that the TTL is applied the way we want: if we insert a new record that _replaces_ and existing one, the expiration time is calculated based on the _new_ record, not the old one.

**Comment by bnaecker** [comment]
One wrinkle here is that setting a TTL relies on there being an actual timestamp somewhere in the table. That's because TTLs are used to expire data as the parts are merged in the background. We don't have any in our current table schema. I think we can get around that pretty easily though, by adding a [`MATERIALIZED` column](https://clickhouse.com/docs/en/sql-reference/statements/create/table#materialized) representing the last time we inserted some data. For example:

```sql
CREATE TABLE IF NOT EXISTS oximeter.fields_uuid
(
    timeseries_name String,
    timeseries_key UInt64,
    field_name String,
    field_value UUID,
    -- New column for TTL only
    last_updated_at DateTime MATERIALIZED now()
)
ENGINE = ReplacingMergeTree()
ORDER BY (timeseries_name, field_name, field_value, timeseries_key)
-- Add a TTL
TTL last_updated_at + INTERVAL 30 DAY;
```

Materialized columns are special. They are calculated automatically at insertion time, and cannot be specified in an `INSERT` query. That means any time we insert new fields, we'll bump this update time. During deduplication, the database will select the [last row with the same sorting key](https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/replacingmergetree#ver), which will be the one with the latest `last_updated_at` timestamp.

Materialized columns also cannot be selected from the tables when doing `SELECT *`. We don't do that for OxQL or other similar operations, but this is still good since it means we won't have to change the database model or serialization at all for this to work. It's entirely inside the DB :)

**Comment by bnaecker** [comment]
Closed by #6352
